fpdf/fpdf.py

```diff
@@ -1266,17 +1267,116 @@ def rect(self, x, y, w, h, style=None):
+            Possible values are:
+
+            *`TOP_LEFT`: a rectangle with round top left corner
+            *`TOP_RIGHT`: a rectangle with round top right corner
+            *`BOTTOM_LEFT`: a rectangle with round bottom left corner
+            *`BOTTOM_RIGHT`: a rectangle with round bottom right corner
+            *`True`: a rectangle with all round corners
+            *`False`: a rectangle with no round corners
             style (fpdf.enums.RenderStyle, str): Optional style of rendering. Possible values are:
 
             * `D` or empty string: draw border. This is the default value.
             * `F`: fill
             * `DF` or `FD`: draw and fill
         """
-        style = RenderStyle.coerce(style)
-        self._out(
-            f"{x * self.k:.2f} {(self.h - y) * self.k:.2f} {w * self.k:.2f} "
-            f"{-h * self.k:.2f} re {style.operator}"
-        )
+
+        if round_corners is not False:
+            self._draw_rounded_rect(x, y, w, h, style, round_corners)
+        else:
+            style = RenderStyle.coerce(style)
+            self._out(
+                f"{x * self.k:.2f} {(self.h - y) * self.k:.2f} {w * self.k:.2f} "
+                f"{-h * self.k:.2f} re {style.operator}"
+            )
+
+    def _draw_rounded_rect(self, x, y, w, h, style, round_corners):
+        min = h
+        r = (w - h) / 2
+
+        if r < 0:
+            r *= -1
+            min = w
+
+        if r == 0:
+            r = w / 5
+
+        if r >= min / 2:
+            r /= min
+
+        point_1 = point_8 = (x, y)
+        point_2 = point_3 = (x + w, y)
+        point_4 = point_5 = (x + w, y + h)
+        point_6 = point_7 = (x, y + h)
+        coor_x = [x, x + w, x, x + w]
+        coor_y = [y, y, y + h, y + h]
+
+        if round_corners is True:
+            round_corners = [
+                Corner.TOP_RIGHT.value,
+                Corner.TOP_LEFT.value,
+                Corner.BOTTOM_RIGHT.value,
+                Corner.BOTTOM_LEFT.value,
+            ]
+
+        if Corner.TOP_RIGHT.value in round_corners:
+            self.arc(coor_x[0], coor_y[0], 2 * r, 180, 270, style=style)
+            point_1 = (x + r, y)
+            point_8 = (x, y + r)
+
+        if Corner.TOP_LEFT.value in round_corners:
+            self.arc(coor_x[1] - 2 * r, coor_y[1], 2 * r, 270, 0, style=style)
+            point_2 = (x + w - r, y)
+            point_3 = (x + w, y + r)
+
+        if Corner.BOTTOM_LEFT.value in round_corners:
+            self.arc(coor_x[3] - 2 * r, coor_y[3] - 2 * r, 2 * r, 0, 90, style=style)
+            point_4 = (x + w, y + h - r)
+            point_5 = (x + w - r, y + h)
+
+        if Corner.BOTTOM_RIGHT.value in round_corners:
+            self.arc(coor_x[2], coor_y[2] - 2 * r, 2 * r, 90, 180, style=style)
+            point_6 = (x + r, y + h)
+            point_7 = (x, y + h - r)
+
+        original_color = self.draw_color.colors
```

Comment:
I'm sorry but I really don't understand why you have to alter the _draw color_...
Could you please explain a bit why things do not work with the color currently set?
Also, if ever changing the _draw color_ is needed, it should be reverted back to its original value at the end of this method call.

Reply:
The logic behind the color chance is to draw a polygon that fills the rectangle when the style is fill. The problem with that is that there is some weird lines between the arcs. In order to solve that I change the draw color to equal fill color so that weird lines dissapears.

Also, you're right, there was a little mistake in the logic and the first change of color shoud be inside the next if. I fixed that, so it changes the draw color and come back to the original color in the same if

Reply:
> The logic behind the color chance is to draw a polygon that fills the rectangle when the style is fill. The problem with that is that there is some weird lines between the arcs. In order to solve that I change the draw color to equal fill color so that weird lines dissapears.

Just draw the polygon with fill only.
With your recent change to `.is_fill` and `.is_draw`, the horizontal and vertical lines are already drawn seperately as appropriate.

@Lucas-C , Independently of this particular PR.: Would it make things simpler or more complicated to have all the shape methods use the drawing module instead of fiddling with these things locally?

Reply:
> Would it make things simpler or more complicated to have all the shape methods use the drawing module instead of fiddling with these things locally?

In terms of "pure" software engineering, yes it would be "cleaner" to use the [drawing](https://pyfpdf.github.io/fpdf2/Drawing.html) instead of directly writing to the internal byte buffer with `FPDF._out`.
But for me this is not a strict requirement as currently most of `FPDF` methods use the 2nd approach. Especially in the case of this PR where the goal is to enhance an exhisting method. A refactor could be done later on, to rewrite some methods to use the `drawing` module, if need be.